adr/ADR-42.md

```diff
@@ -0,0 +1,165 @@
+data.
+
+This is only supported on Pull Consumers, configuring this on a Push consumer must raise an error.
+
+## Priority Policies
+
+### `overflow` policy
+
+Users want certain clients to only get messages when certain criteria are met.
+
+Imagine jobs are best processed locally in `us-east-1` but at times there might be so many jobs in that region that 
+`us-west-1` region handling some overflow, while less optimal in terms of transit costs and latency, would be desirable 
+to ensure serving client needs as soon as possible.
+
+The `overflow` policy enables Pull requests to be served only if criteria like `num_pending` and `num_ack_pending` are
+above a certain limit, otherwise those Pull requests will sit idle in the same way that they would if no messages were
+available (receiving heartbeats etc).
+
+```go
+{
+	PriorityGroups: ["jobs"],
+	PriorityPolicy: "overflow",
+	AckPolicy: "explicit",
+	// ... other consumer options
+}
+```
+
+Here we state that the Consumer has one group called `jobs`, it is operating on the `overflow` policy and requires `explicit`
+Acks, any other Ack policy will produce an error. If we force this ack policy in normal use we should error in Pedantic mode.
+
+Pull requests will have the following additional fields:
+
+ * `"group": "jobs"` - the group the pull belongs to, pulls not part of a valid group will result in an error
+ * `"min_pending": 1000` - only deliver messages when `num_pending` for the consumer is >= 1000
+ * `"min_ack_pending: 1000` - only deliver messages when `ack_pending` for the consumer is >= 1000
+
+If `min_pending` and `min_ack_pending` are both given either being satisfied will result in delivery (boolean OR).
+
+In the specific case where MaxAckPending is 1 and a pull is made using `min_pending: 1` this should only be served when
+there are no other pulls waiting. This means we have to give priority to pulls without conditions over those with when
+considering the next pull that will receive a message.
+
+### `pinned_client` policy
+
+Users want to have a single client perform all the work for a consumer, but they also want to have a stand-by client that
+can take over when the primary, aka `pinned` client, fails.
+
+**NOTE: We should not describe this in terms of exclusivity as there is no such guarantee, there will be times when one
+client think it is pinned when it is not because the server switched.** 
+
+The `pinned_client` policy provides server-side orchestration for the selection of the pinned client.
+
+```go
+{
+	PriorityGroups: ["jobs"],
+	PriorityPolicy: "pinned_client",
+	PriorityTimeout: 120*time.Second,
+	AckPolicy: "explicit",
+	// ... other consumer options
+}
+```
+
+This configuration states:
+
+ * We have 1 group defined and all pulls have to belong to this group
+ * The policy is `pinned_client` that activates these behaviors
+ * When a pinned client has not done any pulls in the last 120 seconds the server will switch to another client
```

Comment:
With this approach, longer expire interval, more sensitive the TTL becomes.
Maybe we should base TTL on Duration since last Pinned Client Pull Request expiry?

Reply:
FWIW in my initial consumer group implementation I am using the idle timeout for the consumer (i.e. time since last pull request expired) and it's working for me. Otherwise what if you set the timeout to for example 60s and the applications are fetching with a timeout of more than 60s, wouldn't that lead to flapping when there are no new messages to deliver?

Reply:
Maybe not clear here but like the way it calculates if a consumer is idle or not it wouldn’t consider a client with an flight pull as being active. 

But this seems open to bugs or abuse so perhaps you simply would have to ensure you pull more frequently. Alternatively we could cap the pull timeout so it relates to this timeout - no more than twice this timeout pull expire or something?

good point. Needs some thinking here